### ISSUE TITLE
FIPS: Failures in SAML tests

### ISSUE THREAD

**Issue by mposolda** [keycloak/keycloak#14969]
When running testsuite with Keycloak on Quarkus distribution with FIPS support and BCFIPS security providers (after #14886), there are failures like those below. There are really lots of failures and hence it should be investigated whether to divide this issue into more issues.

Some snippets from test failures:
```
[ERROR]   SAMLServletAdapterTest.salesPostEncRejectConsent:876 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Referrer-Policy: no-referrer, X-Frame-Options: SAMEORIGIN, Strict-Transport-Security: max-age=31536000; includeSubDomains, X-Content-Type-Options: nosniff, X-XSS-Protection: 1; mode=block, Content-Type: application/json, content-length: 25] ResponseEntityProxy{[Content-Type: application/json,Content-Length: 25,Chunked: false]}}> with entity {"error":"unknown_error"}
[ERROR]   SAMLServletAdapterTest.salesPostSigChangeContents:1033 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Referrer-Policy: no-referrer, X-Frame-Options: SAMEORIGIN, Strict-Transport-Security: max-age=31536000; includeSubDomains, X-Content-Type-Options: nosniff, X-XSS-Protection: 1; mode=block, Content-Type: application/json, content-length: 25] ResponseEntityProxy{[Content-Type: application/json,Content-Length: 25,Chunked: false]}}> with entity {"error":"unknown_error"}
[ERROR]   SAMLServletAdapterTest.salesPostSigStaxParsingFlawEmailTest:1012 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Referrer-Policy: no-referrer, X-Frame-Options: SAMEORIGIN, Strict-Transport-Security: max-age=31536000; includeSubDomains, X-Content-Type-Options: nosniff, X-XSS-Protection: 1; mode=block, Content-Type: application/json, content-length: 25] ResponseEntityProxy{[Content-Type: application/json,Content-Length: 25,Chunked: false]}}> with entity {"error":"unknown_error"}

[ERROR]   SamlSignatureTest.testNoChange:330->testSamlResponseModifications:272->testSamlResponseModificationsClient:292 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Referrer-Policy: no-referrer, X-Frame-Options: SAMEORIGIN, Strict-Transport-Security: max-age=31536000; includeSubDomains, Cache-Control: no-store, must-revalidate, max-age=0, X-Content-Type-Options: nosniff, Set-Cookie: KEYCLOAK_LOCALE=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/demo/; Secure; HttpOnly, Set-Cookie: KC_RESTART=; Version=1; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/demo/; Secure; HttpOnly, Set-Cookie:

[ERROR]   SamlSignatureTest.testXSW5:360->testSamlResponseModifications:272->testSamlResponseModificationsClient:292 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Referrer-Policy: no-referrer, X-Frame-Options: SAMEORIGIN, Strict-Transport-Security: max-age=31536000; includeSubDomains, Cache-Control: no-store, must-revalidate, max-age=0, X-Content-Type-Options: nosniff, Set-Cookie: KEYCLOAK_LOCALE=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/demo/; Secure; HttpOnly, Set-Cookie: KC_RESTART=; Version=1; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/demo/; Secure; HttpOnly, Set-Cookie: KEYCLOAK_IDENTITY=eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJjNWJlOTU0MS0yZDcyLTQ4ZDgtYTZjOC1lMDY0OGYzZTFkNjcifQ.eyJleHAiOjE2NjU1NTAyNzUsImlhdCI6MTY2NTUxNDI3NSwianRpIjoiMDIwYTBjN2ItMTlmNy00MzEwLWJiNjctNjM3YjViZWViZTQ1IiwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6ODU0My9hdXRoL3JlYWxtcy9kZW1vIiwic3ViIjoiYWJmMjViZGEtMDgzMC00MzQzLWE5MzYtNThlNWI1NDAyOWZjIiwidHlwIjoiU2VyaWFsaXplZC1JRCIsInNlc3Npb25fc3RhdGUiOiJlYjc1YTE4YS02NzdiLTQ5MzUtOTczNy0wMjNmNmI2YzRhMzYiLCJzaWQiOiJlYjc1YTE4YS02NzdiLTQ5MzUtOTczNy0wMjNmNmI2YzRhMzYiLCJzdGF0ZV9jaGVja2VyIjoiUmxjMENDRWxhZ05HRnFhTkpjTnY2dTVJRjd0TGljSU1vbWNDQU11Y3A4ayJ9.Kh1Z5BjIri5BWskcNSHsQZ9z0lzwZNVC2Np3ep1kNf4; Version=1; Path=/auth/realms/demo/; SameSite=None; Secure; HttpOnly, Set-Cookie: KEYCLOAK_IDENTITY_LEGACY=eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJjNWJlOTU0MS0yZDcyLTQ4ZDgtYTZjOC1lMDY0OGYzZTFkNjcifQ.eyJleHAiOjE2NjU1NTAyNzUsImlhdCI6MTY2NTUxNDI3NSwianRpIjoiMDIwYTBjN2ItMTlmNy00MzEwLWJiNjctNjM3YjViZWViZTQ1IiwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6ODU0My9hdXRoL3JlYWxtcy9kZW1vIiwic3ViIjoiYWJmMjViZGEtMDgzMC00MzQzLWE5MzYtNThlNWI1NDAyOWZjIiwidHlwIjoiU2VyaWFsaXplZC1JRCIsInNlc3Npb25fc3RhdGUiOiJlYjc1YTE4YS02NzdiLTQ5MzUtOTczNy0wMjNmNmI2YzRhMzYiLCJzaWQiOiJlYjc1YTE4YS02NzdiLTQ5MzUtOTczNy0wMjNmNmI2YzRhMzYiLCJzdGF0ZV9jaGVja2VyIjoiUmxjMENDRWxhZ05HRnFhTkpjTnY2dTVJRjd0TGljSU1vbWNDQU11Y3A4ayJ9.Kh1Z5BjIri5BWskcNSHsQZ9z0lzwZNVC2Np3ep1kNf4; Version=1; Path=/auth/realms/demo/; Secure; HttpOnly, Set-Cookie: KEYCLOAK_SESSION=demo/abf25bda-0830-4343-a936-58e5b54029fc/eb75a18a-677b-4935-9737-023f6b6c4a36; Version=1; Expires=Wed, 12-Oct-2022 04:51:15 GMT; Max-Age=36000; Path=/auth/realms/demo/; SameSite=None; Secure, Set-Cookie: KEYCLOAK_SESSION_LEGACY=demo/abf25bda-0830-4343-a936-58e5b54029fc/eb75a18a-677b-4935-9737-023f6b6c4a36; Version=1; Expires=Wed, 12-Oct-2022 04:51:15 GMT; Max-Age=36000; Path=/auth/realms/demo/; Secure, Set-Cookie: KEYCLOAK_REMEMBER_ME=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/demo/; Secure; HttpOnly, P3P: CP="This is not a P3P policy!", X-XSS-Protection: 1; mode=block, Content-Type: application/json, content-length: 25] ResponseEntityProxy{[Content-Type: application/json,Content-Length: 25,Chunked: false]}}> with entity {"error":"unknown_error"}
[ERROR]   SamlSignatureTest.testXSW6:365->testSamlResponseModifications:272->testSamlResponseModificationsClient:292 
Expected: response status code matches is <200>

[ERROR]   IdentityProviderTest.testSamlExportSignatureOn:1069 expected:<200> but was:<500>

[ERROR]   KcSamlBrokerAllowedClockSkewTest.loginClientExpiredResponseFromIdPWithClockSkew:101 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Referrer-Policy: no-referrer, X-Frame-Options: SAMEORIGIN, Strict-Transport-Security: max-age=31536000; includeSubDomains, X-Content-Type-Options: nosniff, Set-Cookie: KEYCLOAK_LOCALE=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/consumer/; HttpOnly, Set-Cookie: KC_RESTART=; Version=1; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/consumer/; HttpOnly, Set-Cookie: ...

[ERROR]   KcSamlSignedBrokerTest.testSignatureTampering_signDoc_signAssert_encAssert:352->loginAttackChangeSignature:387->loginAttackChangeSignature:432->withSignedEncryptedAssertions:90->lambda$loginAttackChangeSignature$3:447 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Referrer-Policy: no-referrer, X-Frame-Options: SAMEORIGIN, Strict-Transport-Security: max-age=31536000; includeSubDomains, Cache-Control: no-store, must-revalidate, max-age=0, X-Content-Type-Options: nosniff, Set-Cookie: KEYCLOAK_LOCALE=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/provider/; HttpOnly, Set-Cookie: KC_RESTART=; Version=1; ...

[ERROR]   RealmTest.convertSAMLClientDescription:621 » InternalServerError HTTP 500 Internal Server Error
[ERROR]   KcSamlBrokerFrontendUrlTest.testFrontendUrlInDestinationExpected » Runtime java.lang.AssertionError: 
Expected: response status code matches is <200>
     but: was <HttpResponseProxy{HTTP/1.1 500 Internal Server Error [Set-Cookie: KEYCLOAK_LOCALE=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/realms/consumer/; HttpOnly, Set-Cookie: KC_RESTART=; Version=1; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/realms/consumer/; HttpOnly, Set-Cookie: ...
...

```

**Comment by mposolda** [comment]
Investigation: There is similar issue like for Kerberos (https://github.com/keycloak/keycloak/issues/14966) . SAML requires security provider `XMLDSig` to be present in the `java.security` file (corresponding java class is `org.jcp.xml.dsig.internal.dom.XMLDSigRI`). 

This security provider is not available in the `java.security` file by default in the RHEL 8 with FIPS enabled OpenJDK 11. It is to be investigated whether we can just add it or not...

**Comment by mposolda** [comment]
For the reference:
Support for XMLDSig provider in OpenJDK is in the backlog of OpenJDK team with a low priority. Details in the BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1940064
